v1.go

```diff
@@ -97,16 +99,90 @@ func v1(w http.ResponseWriter, r *http.Request) {
 
 			if !r["success"].(bool) {
 				log.Println("Failed to send verification code.")
-				conn.WriteJSON(Action{
-					map[string]interface{}{},
-					"VERIFICATION_VERIFICATION_CODE_SENT_FAILED",
-				})
+				writeEmptyAction(conn, verificationRequestCodeFailure)
 				break
 			}
 
+			clients[conn] = &Client{CountryCode: countryCode, PhoneNumber: phoneNumber}
 			conn.WriteJSON(Action{
 				map[string]interface{}{},
-				"VERIFICATION_VERIFICATION_CODE_SENT",
+				verificationRequestCodeSuccess,
+			})
+		case verificationSubmitCodeRequest:
+			code, ok := action.Payload["code"].(string)
+
+			if !ok {
+				log.Println("code is required in Payload.")
+				writeEmptyAction(conn, verificationSubmitCodeFailure)
+				break
+			}
+
+			client, ok := clients[conn]
+
+			if !ok {
+				log.Println("Unknown client.")
+				writeEmptyAction(conn, verificationSubmitCodeFailure)
+				break
+			}
+
+			if client.CountryCode == "" {
+				log.Println("Client's country code record is missing.")
+				writeEmptyAction(conn, verificationSubmitCodeFailure)
+				break
+			}
+
+			if client.PhoneNumber == "" {
+				log.Println("Client's phone number record is missing.")
+				writeEmptyAction(conn, verificationSubmitCodeFailure)
+				break
+			}
+
+			if client.VerificationCode == "" {
+				url := "https://api.authy.com/protected/json/phones/verification/check"
+				url += "?country_code=" + client.CountryCode
+				url += "&phone_number=" + client.PhoneNumber
+				url += "&verification_code=" + code
+
+				req, err := http.NewRequest("GET", url, nil)
+
+				if err != nil {
+					writeEmptyAction(conn, verificationSubmitCodeFailure)
+					break
+				}
+
+				req.Header.Add("X-Authy-API-Key", twilioAPIKey)
+				resp, err := httpClient.Do(req)
+
+				if err != nil {
+					log.Println("Failed to send verify code.")
+					writeEmptyAction(conn, verificationSubmitCodeFailure)
+					break
+				}
+
+				body, err := ioutil.ReadAll(resp.Body)
+				resp.Body.Close()
+
+				r := map[string]interface{}{}
+				json.Unmarshal(body, &r)
+
+				if !r["success"].(bool) {
+					log.Println("Verification code is incorrect.")
+					writeEmptyAction(conn, verificationRequestCodeFailure)
+					break
+				}
+
+				client.VerificationCode = code
+			} else {
+				if client.VerificationCode != code {
+					log.Println("Verification code doesn't match the record.")
+					writeEmptyAction(conn, verificationRequestCodeFailure)
+					break
+				}
+			}
+
+			conn.WriteJSON(Action{
```

Comment:
@cglotr `writeEmptyAction` here too? Other than that everything looks good to me.

Reply:
Done dc755c3e57b2cbc46428537ca1c48c95680ac168